lib/stitches/auto_fill_test.py

```diff
@@ -0,0 +1,200 @@
+from shapely import geometry as shgeo
+
+from lib.stitches import auto_fill
+from lib.stitches.utils import testing
+from lib.stitch_plan.stitch import Stitch
+
+
+def test_auto_fill_box():
+    # a box with side length 10 and lower-left corner at (10, 10)
+    shape = shgeo.box(10, 10, 20, 20, ccw=True)
+
+    actual = auto_fill(
+        shape=shape,
+        angle=0,
+        row_spacing=1,
+        end_row_spacing=1,
+        max_stitch_length=1,
+        running_stitch_length=1,
+        running_stitch_tolerance=1,
+        staggers=4,
+        skip_last=False,
+        starting_point=(10, 10),
+        ending_point=(10, 10),
+        underpath=False,
+    )
+
+    # the start and ending points should be the same
+    assert actual[0].as_point().isclose(actual[-1].as_point())
+
+    # the exact stitches output by the algorithm now
+    expected = [
+        Stitch(*p).as_point()
+        for p in [
+            (10.0, 10.0),
+            (11.0, 10.0),
+            (12.0, 10.0),
+            (13.0, 10.0),
+            (14.0, 10.0),
+            (15.0, 10.0),
+            (16.0, 10.0),
+            (17.0, 10.0),
+            (18.0, 10.0),
+            (19.0, 10.0),
+            (20.0, 10.0),
+            (20.0, 11.0),
+            (20.0, 12.0),
+            (20.0, 13.0),
+            (20.0, 14.0),
+            (20.0, 15.0),
+            (20.0, 16.0),
+            (20.0, 17.0),
+            (20.0, 18.0),
+            (20.0, 19.0),
+            (20.0, 20.0),
+            (20.0, 20.0),
+            (20.0, 20.0),
```

Comment:
What's going on with these three repeated stitches? There are some other places where a stitch is repeated 2-3 times.

Reply:
I think I've figured it out. The issue is in how the graph path is converted to stitches in `path_to_stitches`.

1. On [this line](https://github.com/inkstitch/inkstitch/blob/main/lib/stitches/auto_fill.py#L630) the initial travel stitch goes from (10, 10) to (20, 20), and the last stitch is on (20, 20).
2. The next edge is a segment edge from (20, 20) to (10, 20). The program goes into `stitch_row` and immediately appends the first stitch (20, 20) to the list [here](https://github.com/inkstitch/inkstitch/blob/main/lib/stitches/fill.py#L77).
3. The `adjust_stagger` just happens to land again on `(20, 20)` for my parameters.

```python
>>> beg
Stitch(20.0, 20.0, None,  ,  ,  ,  ,  ,  ,  )
>>> adjust_stagger(beg, 0, 1, 1, 4)    # my params
Stitch(20.0, 20.0, None,  ,  ,  ,  ,  ,  ,  )

# some other random choices of `staggers`
>>> adjust_stagger(beg, 0, 1, 1, 3)
Stitch(19.666666666666668, 20.0, None,  ,  ,  ,  ,  ,  ,  )
>>> adjust_stagger(beg, 0, 1, 1, 5)
Stitch(20.0, 20.0, None,  ,  ,  ,  ,  ,  ,  )
>>> adjust_stagger(beg, 0, 1, 1, 6)
Stitch(19.333333333333332, 20.0, None,  ,  ,  ,  ,  ,  ,  )
```

My guess is to just skip these appends if the last entry in the stitches list is equal to the thing it's about to append. Does that seem reasonable to you?

Reply:
A fix attempted in cbca04b2e8fdf4a36b5d1ab1b8e7a5127b95ca88, with tests in the following commit

Reply:
Ah, I see now.  Thanks for untangling this!

I wonder if it makes more sense to just leave it to the duplicate stitch filter to remove the extra points rather than having to litter our code with `isclose` calls.  It will also be more efficient, since we already have to calculate the distance between each successive stitch in the duplicate stitch filter.  On the downside, it does make this test look a little weird, having 3 repeated points...

Reply:
Right, the alternative I was thinking of was to write down a list of invariants the function should uphold (you can see that in some of the other tests), and only test those. E.g., the starting/ending points are as specified, travel stitches don't go inside the figure unless the relevant flag is set. But I didn't quite know how to state the most important invariant of all: that the entire region is filled. So I opted for the more specific option.

I was also thinking to try adding some hypothesis tests (generate a random polygon and fill it), which would also benefit from some testable invariants.

Happy for you to make the call as to how this should be resolved. I can run a timing test to show the difference in performance if you'd like. My feeling is it's negligible, so "messy code" would be the deciding factor.